packages/upload/src/vaadin-upload.js

```diff
@@ -795,7 +795,13 @@ class Upload extends ElementMixin(ThemableMixin(ControllerMixin(PolymerElement))
       if (file.xhr) {
         file.xhr.abort();
       }
-      this._notifyFileChanges(file);
+      fileElement.dispatchEvent(
+        new CustomEvent('file-remove', {
+          detail: { file },
+          bubbles: true,
+          composed: true,
+        }),
+      );
```

Comment:
I still find it odd that we dispatch this only to handle it in the same component to then actually remove the file. That is just a more convoluted way of calling `_removeFile`.

How about calling `_removeFile` directly here, and then dispatch the remove event from that method (mainly for legacy purposes / maybe some Flow application listening for it)?

Reply:
OK, makes sense. I guess we should then drop `file-remove` from `vaadin-upload-file` JSDoc. 
As you mentioned, `GeneratedVaadinUploadFile` depends on it, but that's not a public API.